### DISCUSSION TITLE
How do I return an arbitrary status code with an Optional message?

### DISCUSSION THREAD

**Discussion by thomasmost** [rwf2/Rocket#2341] [Questions] [answered]
How do I return a response from an API handler with an arbitrary error status and an arbitrary error message?

I currently make use of response statuses in two ways in my app, with `rocket::http::Status` and `rocket::response::status`. In the former case, I can return any error status I want, but I can't pass down a custom error message with it. See below:

```rs
use rocket::http::Status;

#[get("/hello1")]
pub async fn hello_any_status(
  db_pool: &State<Pool>,
) -> Result<Json<String>, Status> {
  ...
  return Err(Status::BadRequest)
}
```

...and in the latter case, I can pass a custom message, but the only supported error status code is `BadRequest`

```rs
use rocket::response::status;

#[get("/hello2")]
pub fn hello_with_bad_request(
  db_pool: &State<Pool>,
) -> Result<Json<String>, status::BadRequest<String>> {
  ...
  return Err(BadRequest(Some("Your name must be longer than 3 characters")))
}
```

What's the difference between `rocket::http::Status` and `rocket::response::status`, and how can I respond with any appropriate status code AND an optional message?

Thanks in advance 🚀

**Comment by the10thWiz** [comment, marked as answer]
There are a number of options. I recommend deriving [`Responder`](https://api.rocket.rs/v0.5-rc/rocket/derive.Responder.html) on an enum, which allows you to set a custom status & responder. For a simple case, you can return `(http::Status, R)`, where R implements Responder. This will set the status code and return the content of R.

The enum case should looks something like:
```rust
#[derive(Responder)]
pub enum ApiResponse<T> {
  Ok(T),
  #[response(status = 400)]
  BadRequest(String),
  #[response(status = 401)]
  Unauthorized(String),
  #[response(status = 403)]
  Forbidden(String),
}
```

Then, `ApiResponse` can be used in place of `Result`.
